### PR TITLE
Changes so embedded MakeCode is always in the DOM

### PR DESCRIPTION
**Pull request by microbit-robert**
We need this to be able to get hex downloads at any time.

### CODE REVIEW

File: src/user-projects-hooks.tsx [hunk 1]
```diff
@@ -1,128 +1,124 @@
...
+    }
   );
+  const editorWrite: MutableRefObject<
+    ((payload: MakeCodeProject) => void) | undefined
+  > = useRef();
 
-  const setProject = useCallback(
-    (project: MakeCodeProject | undefined) => {
-      setUserProjects({ makeCode: project });
+  const updateProject = useCallback(
```
[the remark this answers is not in the document]
Comment (microbit-robert) [reply src/user-projects-hooks.tsx]: Fixed with debouncing the update [abc94c5](https://github.com/microbit-foundation/ml-trainer/pull/298/commits/abc94c50e97bd69e61c69141855fea5f42bce5b6)

File: src/components/EditCodeDialog.tsx [hunk 2]
```diff
@@ -1,65 +1,138 @@
...
+          if (gestureData) {
+            const parsedGestureData = JSON.parse(
+              gestureData
+            ) as GestureContextState;
+            if (parsedGestureData.lastModified !== gestures.lastModified) {
+              gestureActions.validateAndSetGestures(parsedGestureData.data);
+            }
+          }
+        }
+        // TODO: Test this code after switching to new react-editor-embed version.
```
Comment (microbit-matt-hillsdon) [inline src/components/EditCodeDialog.tsx:49]: TODO

File: src/components/SaveProjectMenuItem.tsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import { FormattedMessage } from "react-intl";
+import { useProject } from "../user-projects-hooks";
+import SaveHexDialog from "./SaveHexDialog";
+
+const SaveProjectMenuItem = () => {
+  const { projectIOState, setProjectIOState } = useProject();
+  const saveHexDisclosure = useDisclosure();
+  const handleDownloadHex = useCallback(() => {
+    saveHexDisclosure.onOpen();
+    setProjectIOState("downloading");
```
Comment (microbit-matt-hillsdon) [inline src/components/SaveProjectMenuItem.tsx:13]: I'd like to understand this state more clearly. When does it get reset? What does it control.

File: src/user-projects-hooks.tsx [hunk 4]
```diff
@@ -1,127 +1,221 @@
...
+  );
 
-  const createGestureDefaultProject = useCallback(
-    (gesture: GestureData) => {
-      const gs = [gesture];
-      return {
+  const debouncedEditorUpdate = useMemo(
+    () =>
+      debounce((project: Project) => {
+        // TODO: consider logging if this fails?
```
Comment (microbit-matt-hillsdon) [inline src/user-projects-hooks.tsx:81]: TODO

File: src/components/SaveProjectMenuItem.tsx (hunk 3 above)
Comment (microbit-matt-hillsdon) [reply src/components/SaveProjectMenuItem.tsx]: Makes sense. Agreed to switch to two dialogs, one with info and don't show this again, one that's just progress. Maybe a toast after the actual download. I'll try to move the "IO" state into the hook.

File: src/components/EditCodeDialog.tsx [hunk 5]
```diff
@@ -1,65 +1,138 @@
...
+    const { project, writeProject, projectIOState, setProjectIOState } =
+      useProject();
+    const { isOpen, onClose } = useEditCodeDialog();
+    const { actions } = useConnectionStage();
+    const [gestures] = useGestureData();
+    const gestureActions = useGestureActions();
+
+    const handleCodeChange = useCallback(
+      (project: Project) => {
+        if (isOpen) {
```
Comment (microbit-matt-hillsdon) [inline src/components/EditCodeDialog.tsx:37]: For my ref: this is the MakeCode hex drop case as well as the block edit case.

**Comment by microbit-matt-hillsdon** [comment]
Closing in favour of the derived #311

**Pull request closed without merging**